website/docs/user-defined-generics.md

```diff
@@ -0,0 +1,163 @@
+---
+id: user-defined-generics
+title: User Defined Generics
+sidebar_label: User Defined Generics
+---
+
+You can define your own generic types, in addition to Sorbet's included
+[standard library generics](stdlib-generics.md) like `T::Array`.
+
+This is done by extending `T::Generic` in your class or module, and then
+creating `type_member` constants for each type parameter that your type needs to
+have.
+
+Here's an example of creating a generic `Box` type, which has one type parameter
+and holds a value of that type:
+
+```ruby
+class Box
+  extend T::Sig
+  extend T::Generic
+
+  # This type has one type member, named "BoxedType"
+  BoxedType = type_member
+
+  sig { params(value: BoxedType).void }
+  def initialize(value)
+    @value = value
+  end
+
+  sig { returns(BoxedType) }
+  attr_accessor :value
+end
+
+# Omitted type members are set to T.untyped
+T.reveal_type(Box.new(3)) # Box[T.untyped]
+
+T.reveal_type(Box[Integer].new(3)) # Box[Integer]
+T.reveal_type(Box[Integer].new(3).value) # Integer
+
+T.reveal_type(Box[Integer].new("hello")) # error: Expected Integer but found String
+```
+
+You can define classes which accept multiple type parameters by using
+`type_member` several times, in order:
+
+```ruby
+class Pair
+  extend T::Sig
+  extend T::Generic
+
+  K = type_member
+  V = type_member
+
+  sig { params(key: K, value: V).void }
+  def initialize(key, value)
+    @key = key
+    @value = value
+  end
+
+  sig { returns(K) }
+  attr_accessor :key
+
+  sig { returns(V) }
+  attr_accessor :value
+end
+
+pair = Pair[String, Integer].new("foo", 2)
```

Comment:
i guess its implied, but it's probably worth noting that the order of parameters here is really significant.

Reply:
I've added some extra comments to show the importance of the order here!